src/transformation/transformation.ts

```diff
@@ -3,9 +3,305 @@ import {
+            });
+
+            factory.finalizer(node);
+            node.parent = parent;
+        }
+        else {
+            if (this.allowGenericNode) {
+                const origin : Origin | undefined = this.asOrigin(source);
+                node = new GenericNode(parent).withOrigin(origin);
+                this._issues.push(
+                    Issue.semantic(
+                        `Source node not mapped: ${Object.getPrototypeOf(source).constructor.name}`,
+                        IssueSeverity.INFO,
+                        origin?.position
+                    )
+                );
+            }
+            else {
+                throw new Error(`Unable to translate node ${source} (class ${Object.getPrototypeOf(source).constructor.name})`)
+            }
+        }
+
+        return node;
+    }
+
+    asOrigin(source: any) : Origin | undefined {
+        if (source instanceof Origin)
+            return source;
+        else
+            return undefined;
+    }
+
+    setChild(
+        childNodeFactory: ChildNodeFactory<any, any, any>,
+        source: any,
+        node: Node,
+        propertyDescription: string
+    ) : void {
+        const src = childNodeFactory.get(this.getSource(node, source));
+
+        let child: any | undefined;
+        if (Array.isArray(src)) {
+            child = src.map(it => this.transform(it, node)).filter(n => n != undefined);
+        }
+        else {
+            child = this.transform(src, node);
+        }
+
+        try {
+            childNodeFactory.set(node, child);
+        } catch (e) {
+            throw new Error(`Could not set child ${childNodeFactory}`);
+        }
+    }
+
+    getSource(node: Node, source: any) : any {
+        return source;
+    }
+
+    makeNode<S extends any, T extends Node>(
+        factory: NodeFactory<S, T>,
+        source: S,
+        allowGenericNode = true
+    ) : Node | undefined {
+
+        let node : Node | undefined;
+
+        try {
+            node = factory.constr(source, this, factory);
+        } catch (e) {
+            if (allowGenericNode)
+                node = new ErrorNode(e);
+            else
+                throw e;
+        }
+
+        if (node)
+            node.withOrigin(this.asOrigin(source));
+
+        return node;
+    }
+
+    getNodeFactory<S extends any, T extends Node>(type: any) : NodeFactory<S, T> | undefined {
+
+        let nodeClass = type;
+
+        if (!isClassType(type) && typeof type === 'object')
+            nodeClass = Object.getPrototypeOf(type).constructor;
+
+        const factory = this.factories.get(nodeClass) as NodeFactory<any, any>;
+        
+        if (factory)
+            return factory as NodeFactory<S, T>;
+
+        // FIXME: The closer ancestor should be given precedence over the farther ones
+        // This is not guaranteed by just cycling over these entries
+        for (const [key, value] of this.factories.entries()) {
+            if (isSuperClass(key, nodeClass))
```

Comment:
What if we count the number of times we have to invoke Object.getPrototypeOf() before arriving at an object p with p.constructor == nodeClass? And we could stop when the prototype is null or undefined, and in that case the count is invalid.

The factory with the lowest count is the one to choose.

Reply:
Ok, I am just cycling over calls of Object.getPrototypeOf(), returning the first match, and it seems to be working now.

Reply:
Cool.